Add ability to override test prefix

schalkwijk opened the pull request:
See [this XCTestList PR](https://github.com/lyndsey-ferguson/xctest_list/pull/7) for more context.

<!-- Thanks for contributing to _test_center_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rake` from the root directory to see all new and existing tests pass
- [x] I've read the [Contribution Guidelines][contributing doc]
- [x] I've updated the documentation if necessary.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
Some testing frameworks for Swift / Objective-C use different test prefixes to identify tests. For instance, [Quick](https://github.com/Quick/Quick) uses the `spec` method name prefix instead of `test`. Thus, we need to be able to identify these differently-named tests, and so this commit introduces the ability to override the prefix that used find tests.
<!-- If it fixes an open issue, please link to the issue here. -->
<!-- Please describe in detail how you tested your changes. -->
I tested my changes by running `multi_scan` with and without my changes against my existing codebase. If I invoke `multi_scan` *without* `test_prefix` (e.g. `bundle exec fastlane run multi_scan try_count:2`), I get the following errors:

```

+--------------------+---+
|      Test Results      |
+--------------------+---+
| Number of tests    | 0 |
| Number of failures | 0 |
+--------------------+---+

[09:46:11]: No tests found in '/Users/user/Library/Developer/Xcode/DerivedData/Application-hbhoorhaiuedoahgciysvkdsihts/Build/Products/Debug-iphonesimulator/Application.app/PlugIns/ApplicationCommonUITests.xctest'!
[09:46:11]: Is the Build Setting, `ENABLE_TESTABILITY` enabled for the test target ApplicationCommonUITests? If your test method names use a prefix other than `test`, consider setting `test_prefix`.
[09:46:11]: No tests found in '/Users/user/Library/Developer/Xcode/DerivedData/Application-hbhoorhaiuedoahgciysvkdsihts/Build/Products/Debug-iphonesimulator/Application.app/PlugIns/ApplicationCoreTests.xctest'!
[09:46:11]: Is the Build Setting, `ENABLE_TESTABILITY` enabled for the test target ApplicationCoreTests? If your test method names use a prefix other than `test`, consider setting `test_prefix`.
[09:46:11]: No tests found in '/Users/user/Library/Developer/Xcode/DerivedData/Application-hbhoorhaiuedoahgciysvkdsihts/Build/Products/Debug-iphonesimulator/Application.app/PlugIns/ApplicationTests.xctest'!
[09:46:11]: Is the Build Setting, `ENABLE_TESTABILITY` enabled for the test target ApplicationTests? If your test method names use a prefix other than `test`, consider setting `test_prefix`.
[09:46:11]: No tests found in '/Users/user/Library/Developer/Xcode/DerivedData/Application-hbhoorhaiuedoahgciysvkdsihts/Build/Products/Debug-iphonesimulator/ApplicationUITests-Runner.app/PlugIns/ApplicationUITests.xctest'!
[09:46:11]: Is the Build Setting, `ENABLE_TESTABILITY` enabled for the test target ApplicationUITests? If your test method names use a prefix other than `test`, consider setting `test_prefix`.
[09:46:11]: No tests found in '/Users/user/Library/Developer/Xcode/DerivedData/Application-hbhoorhaiuedoahgciysvkdsihts/Build/Products/Debug-iphonesimulator/Application.app/PlugIns/ApplicationCommonUITests.xctest'!
[09:46:11]: Is the Build Setting, `ENABLE_TESTABILITY` enabled for the test target ApplicationCommonUITests? If your test method names use a prefix other than `test`, consider setting `test_prefix`.
[09:46:11]: No tests found in '/Users/user/Library/Developer/Xcode/DerivedData/Application-hbhoorhaiuedoahgciysvkdsihts/Build/Products/Debug-iphonesimulator/Application.app/PlugIns/ApplicationCoreTests.xctest'!
[09:46:11]: Is the Build Setting, `ENABLE_TESTABILITY` enabled for the test target ApplicationCoreTests? If your test method names use a prefix other than `test`, consider setting `test_prefix`.
[09:46:12]: No tests found in '/Users/user/Library/Developer/Xcode/DerivedData/Application-hbhoorhaiuedoahgciysvkdsihts/Build/Products/Debug-iphonesimulator/Application.app/PlugIns/ApplicationTests.xctest'!
[09:46:12]: Is the Build Setting, `ENABLE_TESTABILITY` enabled for the test target ApplicationTests? If your test method names use a prefix other than `test`, consider setting `test_prefix`.
[09:46:12]: No tests found in '/Users/user/Library/Developer/Xcode/DerivedData/Application-hbhoorhaiuedoahgciysvkdsihts/Build/Products/Debug-iphonesimulator/ApplicationUITests-Runner.app/PlugIns/ApplicationUITests.xctest'!
[09:46:12]: Is the Build Setting, `ENABLE_TESTABILITY` enabled for the test target ApplicationUITests? If your test method names use a prefix other than `test`, consider setting `test_prefix`.
```

This is because my tests take on the following shape:

```swift
import SwiftUI
import Quick
import Nimble

class OutdatedAppUITest: AbstractUITest {
  var user: TestAPIUser!

  // MAIN SPEC METHOD DOES NOT HAVE `test` PREFIX
  override func spec() {
    beforeSuite {
      self.user = TestAPIClient().createUser()?.setup_complete
      if self.user == nil {
        fail("Could not create user")
      }
    }

    it("sees an update banner and nothing is clickable") {
      expect(self.app.textFields["Email"].isEnabled).to(beFalse())
      expect(self.app.secureTextFields["Password"].isEnabled).to(beFalse())
      expect(self.app.buttons["Sign In"].isEnabled).to(beFalse())

      let bannerTitle = self.app.staticTexts["Get the Latest App"]
      self.waitForExistence(element: bannerTitle, timeout: 15)
      self.tap("Update in the app store")
    }
  }
}
```

However, if I run `bundle exec fastlane run multi_scan try_count:2 test_prefix:spec` (note the `test_prefix` arg), I see proper test output:

```
[09:52:50]: Starting test run 1
[09:52:51]: Starting scan #1 with 5 tests for batch #1.

+------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------+
|                                                                               Summary for scan 2.152.0                                                                                |
+------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------+
| scheme                                         | Application                                                                                                                                |
| only_testing                                   | ["ApplicationUITests/ForgotPasswordUITests/spec", "ApplicationUITests/SecureFieldWithToggleUITests/spec", "ApplicationUITests/SessionUITests/spec",    |
|                                                | "ApplicationUITests/OutdatedAppUITest/spec", "ApplicationUITests/RequestAccountTests/spec"]                                                      |
| output_types                                   | html,junit                                                                                                                           |
| workspace                                      | ./Application.xcworkspace                                                                                                                  |
| skip_detect_devices                            | false                                                                                                                                |
| force_quit_simulator                           | false                                                                                                                                |
| reset_simulator                                | false                                                                                                                                |
| disable_slide_to_type                          | true                                                                                                                                 |
| reinstall_app                                  | false                                                                                                                                |
| app_identifier                                 | co.application.Application                                                                                                                       |
| xctestrun                                      | /Users/user/Library/Developer/Xcode/DerivedData/Application-hbhoorhaiuedoahgciysvkdsihts/Build/Products/Application_iphonesimulator13.5-x86  |
|                                                | _64.xctestrun                                                                                                                        |
| clean                                          | false                                                                                                                                |
| open_report                                    | false                                                                                                                                |
| output_directory                               | /Users/user/application/ios-app/Application/fastlane/test_output/ApplicationUITests-batch-1                                                        |
| output_files                                   | report.html,report.junit                                                                                                             |
| buildlog_path                                  | ~/Library/Logs/scan                                                                                                                  |
| derived_data_path                              | /Users/user/Library/Developer/Xcode/DerivedData/Application-hbhoorhaiuedoahgciysvkdsihts                                               |
| should_zip_build_products                      | false                                                                                                                                |
| use_clang_report_name                          | false                                                                                                                                |
| disable_concurrent_testing                     | true                                                                                                                                 |
| skip_build                                     | false                                                                                                                                |
| xcargs                                         |  -parallel-testing-enabled NO                                                                                                        |
| slack_use_webhook_configured_username_and_icon | false                                                                                                                                |
| slack_username                                 | fastlane                                                                                                                             |
| slack_icon_url                                 | https://fastlane.tools/assets/img/fastlane_icon.png                                                                                  |
| skip_slack                                     | false                                                                                                                                |
| slack_only_on_failure                          | false                                                                                                                                |
| xcodebuild_command                             | env NSUnbufferedIO=YES xcodebuild                                                                                                    |
| fail_build                                     | true                                                                                                                                 |
| include_simulator_logs                         | false                                                                                                                                |
| xcode_path                                     | /Applications/Xcode.app                                                                                                              |
| result_bundle                                  | false                                                                                                                                |
| build_for_testing                              | false                                                                                                                                |
+------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------+

```


### Description
<!-- Describe your changes in detail -->
The changes are fairly simple - we add a single new `test_prefix` argument to the xctestrun fastlane action and use that argument to inform `XCTestList`'s `test` method as to which tests it should find. The bulk of the changes that allow us to find the right tests given the prefix are [here](https://github.com/lyndsey-ferguson/xctest_list/pull/7), which introduces the `test_prefix` keyword arg to `XCTestList`.

lyndsey-ferguson commented on lib/fastlane/plugin/test_center/actions/tests_from_xctestrun.rb:
```diff
@@ -114,6 +114,12 @@ def self.available_options
             default_value: false,
             optional: true
+          ),
+          FastlaneCore::ConfigItem.new(
```
You'll also want to add this new option to [`multi_scan`](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/lib/fastlane/plugin/test_center/actions/multi_scan.rb#L346) so that users of it can pass it through

schalkwijk replied:
Done!

schalkwijk commented:
@lyndsey-ferguson, this is ready for a full, proper review! Tested it locally against my `Quick`-enabled tests and it works wonderfully.

schalkwijk commented on fastlane-plugin-test_center.gemspec:
```diff
@@ -1,3 +1,4 @@
+# coding: utf-8
```
Some editor shenanigans - I'll remove these.

lyndsey-ferguson commented:
I may be able to review this tonight or tomorrow morning. 
#🤞

schalkwijk commented:
Awesome! I look forward to it.

lyndsey-ferguson commented on lib/fastlane/plugin/test_center/actions/multi_scan.rb:
```diff
@@ -357,6 +357,12 @@ def self.available_options
           ),
+          FastlaneCore::ConfigItem.new(
+            key: :test_prefix,
+            description: "The string prefix used to find test methods. Set this to `spec` if you are using Quick",
```
Can you add a `verify_block` to ensure that this is not nil nor an empty string? See [this](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/lib/fastlane/plugin/test_center/actions/multi_scan.rb#L335) as an example for how they work.

lyndsey-ferguson commented on lib/fastlane/plugin/test_center/actions/tests_from_xctestrun.rb:
```diff
@@ -49,7 +49,7 @@ def self.xctestrun_tests(xctestrun_path, invocation_based_tests)
           if test_identifiers.empty? && !invocation_based_tests
             UI.error("No tests found in '#{xctest_path}'!")
-            UI.important("Is the Build Setting, `ENABLE_TESTABILITY` enabled for the test target #{testable_name}?")
+            UI.important("Is the Build Setting, `ENABLE_TESTABILITY` enabled for the test target #{testable_name}? If your test method names use a prefix other than `test`, consider setting `test_prefix`.")
```
```suggestion
            UI.important("Is the Build Setting, `ENABLE_TESTABILITY` enabled for the test target #{testable_name}?")
            UI.message("If your test method names use a prefix other than `test`, consider setting `:test_prefix`.")
```

lyndsey-ferguson commented on lib/fastlane/plugin/test_center/actions/tests_from_xctestrun.rb:
```diff
@@ -114,6 +114,12 @@ def self.available_options
             default_value: false,
             optional: true
+          ),
+          FastlaneCore::ConfigItem.new(
```
See the [comments](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/pull/258/files#diff-b1627ddcb767b566591f53bc2ae1fcffR362) I made about this option in `multi_scan`.

lyndsey-ferguson commented on spec/tests_from_xctestrun_spec.rb:
```diff
@@ -140,4 +140,31 @@
+      allow(File).to receive(:exist?).and_call_original
+      allow(File).to receive(:exist?).with('path/to/fake.xctestrun').and_return(true)
+      allow(File).to receive(:read).with('path/to/fake.xctestrun').and_return(File.read('./spec/fixtures/format-version-2.xctestrun'))
+      expect(XCTestList)
```
What this is doing is "pretending to be the `xctest_list` class that receives the message, "hey, what tests do you have for this prefix?". You have it returning tests that are prefixed with test and stating that you expect to be called and return these mocked tests.

The expectation is fine, but since we don't really care what is being returned, I don't think that you need to add the:
```ruby
.and_return(
..
)
```
part.

schalkwijk replied:
Unfortunately, it looks like we have to return at least an empty array to make sure the rest of the code can function as expected - is that all right?

lyndsey-ferguson replied:
Oh, true. Yes, that should be fine.

lyndsey-ferguson commented on lib/fastlane/plugin/test_center/helper/test_collector.rb:
```diff
@@ -25,6 +25,8 @@ def initialize(options)
           @batch_count = options[:parallel_testrun_count]
         end
+
+        @test_prefix = options[:test_prefix] || "test"
```
If you put in the validation in the ConfigItem in `multi_scan`, do you need this? I don't mind it: I am asking you and the Universe (my subconscious lives there).

schalkwijk replied:
You're totally right - this is redundant and can be removed.

schalkwijk commented:
@lyndsey-ferguson, I believe all comments have been addressed!

schalkwijk commented:
(although there seems to be an issue with Github's interface updating with my commits - I guess we'll have to wait for that eventual concurrency, eh?)

schalkwijk commented:
@lyndsey-ferguson, we should be good now!

lyndsey-ferguson commented:
Thanks! I will be able to review and test tomorrow and over the weekend. If all is good I imagine that this will release next early next week.

schalkwijk commented:
Amazing - keep me posted if anything isn't up to spec.

lyndsey-ferguson commented:
@schalkwijk one thing, can you include a sample project that uses the Quick framework to test? I'd like to have that in order to test manually, and for future testing to ensure that the feature doesn't get broken by other changes.

schalkwijk commented:
@lyndsey-ferguson, added! I basically just copy-pasted the existing `AtomicBoy` project, added Quick to the Podfile, deleted everything but a couple of test files, re-wrote those test files to use Swift + Quick, and verified that they run in XCode as expected. 

Also, in doing so, I realized that my solution to this problem fixes the very narrow case of Quick + Swift, but *not* Quick + Objective-C, which uses a different form entirely to do tests:

```objective-c
#import <Quick/Quick.h>
#import <Nimble/Nimble.h>

QuickSpecBegin(DolphinSpec)

describe(@"a dolphin", ^{
  __block Dolphin *dolphin = nil;
  beforeEach(^{
      dolphin = [Dolphin new];
  });

  describe(@"its click", ^{
    __block Click *click = nil;
    beforeEach(^{
      click = [dolphin click];
    });

    it(@"is loud", ^{
      expect(@(click.isLoud)).to(beTruthy());
    });

    it(@"has a high frequency", ^{
      expect(@(click.hasHighFrequency)).to(beTruthy());
    });
  });
});

QuickSpecEnd
```

Note that there are no function declarations here to piggy-back off of, so I wonder if we should write some clarifying comment somewhere about the `test_prefix` being set to `spec` only works for Swift + Quick?

lyndsey-ferguson commented:
Can you attach the project, I don't want to do the same work you have already done. Also, I do want to look at the symbols provided by the ObjC version of the Quick tests to look for patterns. I think we also have to change `xctest_list` for ObjC tests.

If you can provide me the project, I can do these things:
1. Look for the symbols of QuickTests in xctest_list.
2. Brainstorm the right way to handle this type of pattern for ObjC tests.
3. Brainstorm how to proceed.

I have this plugin supporting both ObjC and Swift, and I want to find a way that I can do so, putting a pause on this PR.

schalkwijk commented:
@lyndsey-ferguson, makes perfect sense to me! I've pushed up a set of commits that introduce some objective-c tests using Quick, so I think that might give you enough to go off of, at least in your initial explorations.

Please, tell me if there's anything else I can help with - I think I'll start using my own fork while I wait for this to wrap up, but I would like to help you carry the entire feature over the finish line.

lyndsey-ferguson commented:
I finally have some time to look at this. First step, observe the Quick tests provided by the test binary to see if I can extract a pattern that can be used to send to `-only-testing`.

schalkwijk commented:
@lyndsey-ferguson, thanks for the update!

lyndsey-ferguson commented:
I couldn't find an easy way to get the names of the tests from the ObjC symbols. What I do see, if I follow the ObjC definitions, is the underlying way that methods are created. I can find those symbols in the binary, but it isn't easy to decipher the names of the actual tests as they become the filename and line number of where the test is.

So, here is what I am thinking: ask you to:
1. Update `xctest_list` to make it clearer that the prefix _only_ applies to Swift tests.
2. Update this PR to make it clearer that this prefix _only_ applies to Swift tests.

Maybe additionally, there is something that we can do in `xctest_list` to print out a warning if it detects an ObjC class with that prefix and a block invoke (you can see that by running `nm -U` on the binary).

schalkwijk commented:
@lyndsey-ferguson, sorry for the delay! This has been updated to rename everything to `swift_test_prefix`, and I've added some clarifying descriptions to the documentation to make it clear that this flag only controls Swift test prefixes. Tell me if there's anything else you need from me.

lyndsey-ferguson commented:
Released in [v3.13.0](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/releases/tag/v3.13.0)!

*AND* [🐦 Tweeted!](https://twitter.com/lyndseydf/status/1289967180222668802)
